### PR TITLE
Stateless Code + ReactPHP integration

### PR DESCRIPTION
**Pull request by mmoreram**
- On road to use Promises on the project
- First of all, services MUST be stateless, even the middleware classes.
- We should integrate **all** the code base with a Promises implementation (ReactPHP, Guzzle...)
- We should work as well in an async kernel implementation for Symfony. Making a PSR proposal would be a win
- Once these pieces are done, then we should join them all, and make some wishes.

## Benchmark

For this PR, I've made some benchmarks analysis. For this reason, we have deployed an index of 11343 items and I have enabled several environments with several entrypoints. In all scenarios, I've used the ab command line `ab -c 100 -n 1000 localhost:8500/v1/marvel?token=0e4d75ba-c640-44c1-a745-06ee51db4e93`. Using production environments without debug.

### Apache2

```
Concurrency Level:      100
Time taken for tests:   6.210 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      24297000 bytes
HTML transferred:       24079000 bytes
Requests per second:    161.04 [#/sec] (mean)
Time per request:       620.970 [ms] (mean)
Time per request:       6.210 [ms] (mean, across all concurrent requests)
Transfer rate:          3821.04 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   1.1      0       4
Processing:    44  599 136.3    602    1516
Waiting:       43  590 133.8    595    1515
Total:         48  599 135.5    603    1516

Percentage of the requests served within a certain time (ms)
  50%    603
  66%    627
  75%    647
  80%    658
  90%    692
  95%    742
  98%    779
  99%   1052
 100%   1516 (longest request)
```

### PHP-PM server

```
Concurrency Level:      100
Time taken for tests:   3.584 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      24296000 bytes
HTML transferred:       24079000 bytes
Requests per second:    278.99 [#/sec] (mean)
Time per request:       358.437 [ms] (mean)
Time per request:       3.584 [ms] (mean, across all concurrent requests)
Transfer rate:          6619.46 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0   59 238.9      0    1023
Processing:    10  284 119.3    364     571
Waiting:       10  284 119.3    364     571
Total:         13  343 281.6    371    1593

Percentage of the requests served within a certain time (ms)
  50%    371
  66%    377
  75%    379
  80%    380
  90%    383
  95%   1204
  98%   1525
  99%   1560
 100%   1593 (longest request)
```

### Symfony Async Kernel + ReactPHP Server

```
Concurrency Level:      100
Time taken for tests:   14.432 seconds
Complete requests:      1000
Failed requests:        0
Total transferred:      24296000 bytes
HTML transferred:       24079000 bytes
Requests per second:    69.29 [#/sec] (mean)
Time per request:       1443.154 [ms] (mean)
Time per request:       14.432 [ms] (mean, across all concurrent requests)
Transfer rate:          1644.08 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0   99 300.8      0    1032
Processing:   105  630 1806.8    245   13414
Waiting:      105  630 1806.8    245   13414
Total:        107  729 2016.2    245   14430

Percentage of the requests served within a certain time (ms)
  50%    245
  66%    247
  75%    248
  80%    249
  90%    346
  95%   2930
  98%   7906
  99%  14412
 100%  14430 (longest request)
```

### CODE REVIEW

File: Domain/Middleware/ExceptionsMiddleware.php [hunk 1]
```diff
@@ -56,27 +56,26 @@ public function __construct(EventPublisher $eventPublisher)
...
-                        : RepositoryReference::create(),
-                    new ExceptionWasCached(new StorableException(
-                        $exception->getMessage(),
-                        (int) $exception->getCode(),
-                        $exception->getTraceAsString(),
-                        $exception->getFile(),
-                        (int) $exception->getLine()
+        return $next($command)
+            ->then(null, function ($exception) use ($command) {
+                var_dump($exception->getMessage());
```
Comment (hkulekci) [inline Domain/Middleware/ExceptionsMiddleware.php]: `var_dump` here

Comment (mmoreram) [reply Domain/Middleware/ExceptionsMiddleware.php]: Yes, thanks :dancing_men: Just a WIP! ^^

File: Plugin/Elastica/Domain/AppRepository/IndexRepository.php [hunk 2]
```diff
@@ -20,122 +20,133 @@
...
  */
-class IndexRepository extends ElasticaWrapperWithRepositoryReference implements IndexRepositoryInterface
+class IndexRepository extends WithElasticaWrapper implements IndexRepositoryInterface
 {
+    /**
+     * @var bool
+     *
+     * Async
+     */
+    private $async = false;
```
Comment (hkulekci) [inline Plugin/Elastica/Domain/AppRepository/IndexRepository.php]: unused private class field

File: Plugin/Elastica/Adapter/AsyncMultiSearch.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+     * @return PromiseInterface
+     */
+    public function searchAsync(): PromiseInterface
+    {
+        $data = $this->_getData();
+        $builder = new MultiBuilder();
+
+        return $this
+            ->getClient()
+            ->requestAsync(
```
Comment (hkulekci) [inline Plugin/Elastica/Adapter/AsyncMultiSearch.php]: `getClient` method returns a `\Elastica\Client` and this has not a `requestAsync`method. We need to change `Elastica\Search::getClient()` method return type as `Apisearch\Plugin\Elastica\Adapter\AsyncClient`

File: Plugin/Elastica/Adapter/AsyncClient.php [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+     */
+    public function requestAsync(
+        string $path,
+        string $method = Request::GET,
+        $data = [],
+        array $query = [],
+        $contentType = Request::DEFAULT_CONTENT_TYPE
+    ): PromiseInterface {
+        if (is_array($data)) {
+            $data = json_encode($data);
```
Comment (hkulekci) [inline Plugin/Elastica/Adapter/AsyncClient.php]: We need to add composer.json file `ext-json`

File: Plugin/Elastica/Adapter/AsyncClient.php [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    ): PromiseInterface {
+        $cloned = clone $endpoint;
+        if ($index instanceof Index) {
+            $cloned->setIndex($index->getName());
+        }
+
+        return $this->requestAsync(
+            ltrim($cloned->getURI(), '/'),
+            $cloned->getMethod(),
+            null === $cloned->getBody() ? [] : $cloned->getBody(),
```
Comment (hkulekci) [inline Plugin/Elastica/Adapter/AsyncClient.php]: will `getBody()` works two times. we can use `$cloned->getBody() ?? []` instead.

File: Controller/Listener/RouterRedirectionToJsonResponse.php [hunk 6]
```diff
@@ -29,38 +29,41 @@ class RouterRedirectionToJsonResponse
...
-                        'code' => Response::HTTP_MOVED_PERMANENTLY,
-                        'message' => 'Moved Permanently',
-                    ],
-                    Response::HTTP_MOVED_PERMANENTLY,
-                    $response->headers->all()
-                ));
-            }
-        }
+                        $response->headers->set('location', $location);
+                        $event->setResponse(new JsonResponse(
```
Comment (hkulekci) [inline Controller/Listener/RouterRedirectionToJsonResponse.php]: This line gives me warning like 

```
Method 'setResponse' not found in \Symfony\Component\HttpKernel\Event\FilterResponsePromiseEvent 
Inspection info: Referenced method is not found in subject class.
```

File: Console/AddTokenCommand.php [hunk 7]
```diff
@@ -90,20 +90,18 @@ protected function runCommand(InputInterface $input, OutputInterface $output)
...
+            $this->createGodToken($objects['app_uuid']),
+            new Token(
+                $objects['token_uuid'],
+                $objects['app_uuid'],
+                $objects['indices_uuid'],
+                $endpoints,
+                $input->getOption('plugin'),
+                (int) $input->getOption('ttl')
+            )
+        ));
```
Comment (hkulekci) [inline Console/AddTokenCommand.php]: Also here `runCommand()` method returns something like `mixed|null` according to comment. But returns nothing.

File: Domain/CommandHandler/CleanEnvironmentHandler.php [hunk 8]
```diff
@@ -26,9 +28,11 @@ class CleanEnvironmentHandler
      * Add token.
      *
      * @param CleanEnvironment $cleanEnvironment
+     *
+     * @return PromiseInterface
      */
-    public function handle(CleanEnvironment $cleanEnvironment)
+    public function handle(CleanEnvironment $cleanEnvironment): PromiseInterface
```
Comment (hkulekci) [inline Domain/CommandHandler/CleanEnvironmentHandler.php]: Where we are using `$cleanEnvironment` variable. I could not find any subclass extends by `CleanEnvironmentHandler`

File: Plugin/RabbitMQ/Domain/RabbitMQConsumerManager.php [hunk 9]
```diff
@@ -144,46 +153,55 @@ function (AbstractConnection $connection) use ($type, $data) {
...
+                    function (AbstractConnection $connection) use ($queueName) {
+                        return $connection
+                            ->channel()
+                            ->queue_declare($queueName, true);
+                    },
+                    $this->getConnection()
+                )
+            ))
+            ->then(function ($data) {
+                return \intval($data[1]);
```
Comment (hkulekci) [inline Plugin/RabbitMQ/Domain/RabbitMQConsumerManager.php]: `(int)$data[1]` instead.

Comment (mmoreram) [reply Plugin/RabbitMQ/Domain/RabbitMQConsumerManager.php]: seems that \intval() is much faster than (int)

File: Domain/CommandHandler/CleanEnvironmentHandler.php (hunk 8 above)
Comment (mmoreram) [reply Domain/CommandHandler/CleanEnvironmentHandler.php]: Anywhere. But this is a command handler, and a command handler requires a command. As this is a value object, this object internal implementation could change in the future, and could be required.

File: Controller/Listener/RouterRedirectionToJsonResponse.php (hunk 6 above)
Comment (mmoreram) [reply Controller/Listener/RouterRedirectionToJsonResponse.php]: ups. This is an issue then.

File: Plugin/Elastica/Adapter/AsyncClient.php (hunk 5 above)
Comment (mmoreram) [reply Plugin/Elastica/Adapter/AsyncClient.php]: yes :)

File: Plugin/Elastica/Adapter/AsyncClient.php (hunk 4 above)
Comment (mmoreram) [reply Plugin/Elastica/Adapter/AsyncClient.php]: yes

File: Plugin/Elastica/Domain/AppRepository/IndexRepository.php (hunk 2 above)
Comment (mmoreram) [reply Plugin/Elastica/Domain/AppRepository/IndexRepository.php]: yes

File: Controller/Listener/RouterRedirectionToJsonResponse.php (hunk 6 above)
Comment (hkulekci) [reply Controller/Listener/RouterRedirectionToJsonResponse.php]: So we need to change tests :D

**Comment by hkulekci** [comment]
Ping @mmoreram what will we do for this pr? Any help?

**Comment by mmoreram** [comment]
@hkulekci it needs some extra time. Time for revision and some time for checking other options for reactive php implementation. This was an exercise to prove that is possible. As soon as I make some reviewing at https://www.swoole.co.uk/, at the PHP async extension, and test some other PHP event loop implementations, I will make a decision.

ATM, I'm not happy yet with the performance of the tested event loop implementation (StreamEventLoop)

**Comment by mmoreram** [comment]
@hkulekci checking this branch, and updating with ReactPHP RabbitMQ implementation - https://github.com/jakubkulhan/bunny

File: Controller/GetTokensController.php [hunk 10]
```diff
@@ -31,27 +32,28 @@ class GetTokensController extends ControllerWithBus
...
      * @param Request $request
      *
-     * @return JsonResponse
+     * @return PromiseInterface
      */
-    public function __invoke(Request $request): JsonResponse
+    public function __invoke(Request $request): PromiseInterface
     {
-        $tokens = $this
+        return $tokens = $this
```
Comment (hkulekci) [inline Controller/GetTokensController.php]: No need to assignment

Comment (mmoreram) [reply Controller/GetTokensController.php]: Absolutely :) Thanks!

**Comment by hkulekci** [comment]
LGTM. But what about tests?

**Comment by mmoreram** [comment]
Working on it :)

**Comment by clue** [comment]
@mmoreram Nice work! :muscle:

The benchmark results appear to be a bit strange, i.e. most requests are much faster, but some are much slower, thus resulting in a worse number of requests per second than your Apache setup.

This is most likely the result of your high concurrency setting in your benchmark (`-c 100`) which will exceed the default backlog size of the listening socket. You can simply assign a higher backlog size and you should see a significant improvement as per https://github.com/reactphp/http/issues/311#issuecomment-376590295:

```php
$socket = new \React\Socket\Server('0.0.0.0:8080', $loop, array('backlog' => 128));
```

**Comment by mmoreram** [comment]
@clue Thanks so much! That refactor wouldn't have been possible without you and your team work.

Related to that numbers, I will take a look at that! Even in my Medium series, I noticed that the last 5% of the long tail is always that bad, but take in account that this benchmark is performed with only one server, what means one CPU. To be fair, that should be performed by using as many cores as my computer has in order to be equal efficient that Apache / NGinx is, right?

Thanks for your feedback!

**Comment by clue** [comment]
> […] this benchmark is performed with only one server, what means one CPU. To be fair, that should be performed by using as many cores as my computer has in order to be equal efficient that Apache / NGinx is, right?

Yes, ReactPHP's HTTP server is single threaded by default and depending your use case your may want to scale this by taking advantage of all your available cores by running multiple instances. There are multiple options for distributing load, e.g. using a load balancer. I can't seem to find an issue discussing this, perhaps you care to file one in our HTTP component to discuss this further? :+1:

**Pull request merged**